### PR TITLE
fix(heck-relay): stop overwriting Rosales OPT params with Seminario projections

### LINKED ISSUE #277: Diagnose heck-relay Seminario divergence (bond_length R² ≈ -48, JaxLoss diverges)
## Background

The fresh convergence baseline (#274, ericchansen/q2mm-data#1) shows
heck-relay with **catastrophically poor Seminario fit** even though the
loader starts from the published Rosales force field:

| Category | n_refs | R² | RMSD | MAE |
|---|---|---|---|---|
| bond_length | 1140 | -48.06 | 2.29 | 1.13 |
| bond_angle | 2157 | -5.47 | 43.7 | 31.9 |
| eig_diagonal | 3121 | -4.66 | 0.31 | 0.18 |

The resulting JaxLoss surrogate diverges by ~21 orders of magnitude on
the very first evaluation. This blocks any JaxLoss-guided optimization
and makes finite-difference optimization impractical (462 active
parameters × ~20 s/eval = hours per gradient step).

The previous stale data reported R²_bond = -434 (now -48), suggesting
something changed between the two runs — but the more important
finding is that **even the better number is catastrophically negative**.

## Loader (current behavior)

`q2mm/diagnostics/systems.py:load_heck_relay()`:

```python
ff_template = ForceField.from_mm3_fld(ff_path, include_standard=True)
ff_opt = ForceField.from_mm3_fld(ff_path, include_standard=False)
ff_template.freeze_standard_params(ff_opt)
ff = estimate_force_constants(molecules, forcefield=ff_template,
                              invert_ts_curvature=True)
```

i.e. it loads the Rosales published FF (which includes both standard
MM3 and the Rosales OPT overlay), freezes the published OPT parameters,
and then re-estimates the **standard** parameters via Seminario.

## Bounded diagnosis (proposed)

**Time-box: ~2 h of investigation. If no clear culprit emerges, file a
deeper investigation issue and stop.**

Three baselines, no optimization, just evaluation:

1. **A — Untouched Rosales FF.** `ForceField.from_mm3_fld(ff_path,
   include_standard=True)`. No `freeze_standard_params`, no Seminario
   re-estimation.
2. **B — Current loader.** As above, with Seminario re-estimating only
   standard params with OPT frozen.
3. **C — Seminario-only.** No published OPT at all — pure MM3 backbone
   + Seminario.

For each baseline, compute:

- Per-category R² / RMSD / MAE (bond_length, bond_angle, eig_diagonal)
- Per-molecule R² + worst-10 bond-length residuals
- Initial ObjectiveFunction and (where finite) JaxLoss

**Decision tree:**

- If A's bond_length R² ≫ B's → the bug is in
  `freeze_standard_params` + Seminario interaction.
- If A's bond_length R² ≈ B's ≈ C's → the bug is upstream
  (MM3 evaluator on Heck atom types, or Rosales-style atom types not
  mapping cleanly).
- If A is fine and B explodes → narrow on what `Seminario` does to
  parameters frozen by `freeze_standard_params`.

## Acceptance criteria

- [ ] Diagnostic script committed under `scripts/` (one-shot, not a
      benchmark — won't replace the regen script).
- [ ] Output committed to `q2mm-data/benchmarks/heck-relay/diagnostic/`.
- [ ] One of:
      - (a) Loader/code fix landed in q2mm, with a `regenerate_convergence_results.py`
            re-run showing R² in a non-catastrophic range; OR
      - (b) Issue closed with "no fix in 2-h window — promoted to a
            deeper investigation issue" and the diagnostic data committed.
- [ ] `docs/systems/heck-relay.md` updated with the finding.

## Out of scope

- Rewriting Seminario implementation
- Adding new atom-type machinery
- Anything that touches multiple systems (heck-relay-specific only)

Refs ericchansen/q2mm#274, ericchansen/q2mm-data#1.

### PR DESCRIPTION
**Pull request by ericchansen**
Resolves #277.

## TL;DR

The Heck-relay loader was silently corrupting the published Rosales OPT force field by overwriting its OPT parameters with raw Seminario projections. Fixing this single line drops the ObjectiveFunction by 43×, takes JaxLoss from "diverges by 21 orders of magnitude" to ratio ≈ 1.29, and brings bond_length R² from **−48** to **0.98**.

## The bug

```python
ff_template = ForceField.from_mm3_fld(ff_path, include_standard=True)
ff_opt = ForceField.from_mm3_fld(ff_path, include_standard=False)
ff_template.freeze_standard_params(ff_opt)
ff = estimate_force_constants(molecules, forcefield=ff_template, ...)  # 💥
```

`freeze_standard_params` marks the *non*-OPT params (standard MM3 backbone) as frozen and leaves the OPT params active. `estimate_force_constants` skips frozen params (see `q2mm/models/seminario.py:341`) and re-estimates the *active* ones — overwriting the carefully-fitted Rosales OPT values with raw Seminario projections.

## Three-baseline diagnostic

The new `scripts/diagnose_heck_relay.py` builds three FFs and evaluates each (no optimization, just evaluation). Output committed to [`q2mm-data/benchmarks/heck-relay/diagnostic/`](https://github.com/ericchansen/q2mm-data/blob/main/benchmarks/heck-relay/diagnostic/three_baseline_comparison.json):

| Baseline | obj | ratio | R²(bond_len) | R²(bond_ang) | R²(eig_diag) |
|---|---|---|---|---|---|
| A — Untouched Rosales | 3.42 × 10⁶ | **1.31** | **0.980** | 0.786 | −12.6 |
| B — Current loader | 1.17 × 10⁸ | NaN | −4787 | −7.25 | −4.66 |
| C — Seminario only | 1.24 × 10⁸ | 1.77 × 10⁷³ | −849 | −5.13 | 0.05 |

A's bond_length R² (0.980) is dramatically better than B's (−4787) — matching the issue's decision rule "A fine, B explodes → narrow on Seminario behavior with freeze_standard_params".

## The fix

Keep `freeze_standard_params` (still needed for the optimizer's active mask) but **skip `estimate_force_constants`**. Use the published Rosales OPT parameters as-is.

## Post-fix baseline

(Companion q2mm-data PR regenerates the heck-relay convergence JSONs.)

| Metric | Before (loader bug) | After (this PR) |
|---|---|---|
| Initial ObjectiveFunction | 1.48 × 10⁸ | 3.46 × 10⁶ |
| Initial JaxLoss | 1e30 (penalty) | 4.48 × 10⁶ |
| Ratio | ~10²¹ | 1.29 |
| Status | out_of_band | out_of_band (12 % over band) |
| bond_length R² | −48.06 | 0.980 |
| bond_angle R² | −5.47 | 0.786 |
| eig_diagonal R² | −4.66 | −12.6 |

The eigenmatrix actually gets worse (−4.66 → −12.6), but that just reflects the real MM3* ↔ JAX-engine cross-engine gap that's now visible without being masked by the corrupted FF. Geometry is now well-reproduced and JaxLoss is no longer divergent.

## Out of scope

- The same `freeze_standard_params + estimate_force_constants` pattern exists in `load_rh_enamide` and the `load_pd_*` loaders. They aren't catastrophically broken today (Rh-enamide runs at R² = 0.97, Pd-allyl at −1.41), but the same audit should probably extend to them in a follow-up.
- Heck-relay now joins pd-conjugate as a candidate for the `ratio_tol=None` experiment ([#276](https://github.com/ericchansen/q2mm/issues/276)).

## Validation

- `python -m ruff check q2mm/ test/ scripts/` — clean.
- `python -m ruff format --check q2mm test scripts examples` — clean.
- `python -m pytest test/ -x -q -m "not (openmm or tinker or jax or jax_md or psi4)" -k "heck or systems"` — 11 passed.
- End-to-end `scripts/regenerate_convergence_results.py --system heck-relay --skip-optimization` ran clean on a single GPU; output (with full provenance) is in the companion q2mm-data PR.

## Companion PR

[`ericchansen/q2mm-data#?`](https://github.com/ericchansen/q2mm-data/pulls) — adds the diagnostic JSON and the regenerated convergence baseline.

### CODE REVIEW

File: scripts/diagnose_heck_relay.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    ff.functional_form = FunctionalForm.MM3
+    return ff
+
+
+def build_ff_b_current_loader(molecules: list[Any]) -> Any:
+    """Build baseline B — current ``load_heck_relay()`` loader output."""
+    from q2mm.diagnostics.systems import load_heck_relay
+
+    sys_data = load_heck_relay(engine=None)
+    return sys_data.forcefield
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply scripts/diagnose_heck_relay.py]: Fixed in 4e86326. Renamed `build_ff_b_current_loader` → `build_ff_b_prefix_loader_pattern` and reconstructed the buggy combination inline (`from_mm3_fld(include_standard=True)` + `freeze_standard_params` + `estimate_force_constants(forcefield=ff)`). The diagnostic no longer calls the current loader, so it stays a valid regression-detection tool after the fix. Label in the output JSON renamed accordingly.

File: scripts/diagnose_heck_relay.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+            cats.get("eig_diagonal", {}).get("r2", float("nan")),
+        )
+
+    provenance = {
+        "generator": "scripts/diagnose_heck_relay.py",
+        "timestamp_utc": datetime.now(timezone.utc).isoformat(),
+        "command_line": shlex.join(sys.argv),
+        "q2mm": _git_info(REPO_ROOT),
+        "q2mm_data": _git_info(args.output.parent.parent.parent.parent),
+        "devices": _device_info(),
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply scripts/diagnose_heck_relay.py:387]: Fixed in 4e86326. New `_find_git_repo_root(start)` walks parents from the output path until a `.git` entry is found. Falls back to the output's parent if no repo is found (`_git_info` then returns `{}` cleanly). No longer assumes the default 4-level layout.

File: docs/systems/heck-relay.md [hunk 3]
```diff
@@ -63,37 +63,46 @@ complete failure of cross-engine transfer, not a small miss.
...
-    surrogate is clearly unusable.  **No JaxLoss-guided optimization
-    was performed.**
+!!! success "Loader bug fixed (ericchansen/q2mm#277) — Rosales FF now usable"
+    The previous loader silently re-ran `estimate_force_constants` on
+    the Rosales FF, which overwrote the published OPT parameters with
+    raw Seminario projections.  That dropped bond_length R² to ≈ −48
+    and made JaxLoss diverge by ~21 orders of magnitude.  The loader
+    now keeps the Rosales OPT values as-is (it still calls
+    `freeze_standard_params` so the active mask is correct for
+    optimization, but no Seminario re-estimation).
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply docs/systems/heck-relay.md]: Fixed in 4e86326. The two numbers were both legitimate (−48 from the prior committed convergence baseline; −4787 from the diagnostic baseline B in this PR), but the doc didn't make that clear. The doc now says the bug is non-deterministic (catastrophically negative R² with the magnitude varying across runs), cites the diagnostic value (−4787) as canonical for this PR, and notes the prior baseline (−48) as historical context.

File: docs/systems/heck-relay.md [hunk 4]
```diff
@@ -63,37 +63,46 @@ complete failure of cross-engine transfer, not a small miss.
...
-correctly identifies this and blocks optimization.
-
-Finite-difference gradients would work in principle but are
-impractical for this system (462 active parameters × ~20 s per
-evaluation = hours per gradient step).
+| Ratio check | 1.29 (out_of_band; upper bound 1.15) |
+| Initial ObjectiveFunction score | 3.46 × 10⁶ |
+| Optimization | Skipped at default `ratio_tol=0.15` (borderline — candidate for `ratio_tol=None`) |
+
+Per-category Seminario fit (now using the untouched Rosales FF):
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply docs/systems/heck-relay.md]: Fixed in 4e86326. Relabeled the table as "Per-category fit of the published Rosales force field against the QM training data (no QFUERZA — these are the published OPT values evaluated by the q2mm JAX engine)". No more contradiction.

File: q2mm/diagnostics/systems.py [hunk 5]
```diff
@@ -354,10 +361,13 @@ def load_heck_relay(
...
-    ff_template = ForceField.from_mm3_fld(str(ff_path), include_standard=True)
+    ff = ForceField.from_mm3_fld(str(ff_path), include_standard=True)
+    # Mark only the OPT-substructure parameters as active for downstream
+    # optimization; standard MM3 backbone params remain frozen.  Do NOT
+    # call estimate_force_constants here — see #277 (it would overwrite
+    # the Rosales-fitted OPT values with raw Seminario projections).
     ff_opt = ForceField.from_mm3_fld(str(ff_path), include_standard=False)
-    ff_template.freeze_standard_params(ff_opt)
-    ff = estimate_force_constants(molecules, forcefield=ff_template, invert_ts_curvature=True)
+    ff.freeze_standard_params(ff_opt)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/diagnostics/systems.py:370]: Fixed in 4e86326. Added `test_load_heck_relay_preserves_published_opt_values` in `test/integration/test_heck_validation.py` (marked `@pytest.mark.validation` + `@pytest.mark.external_data` per the existing pattern). It calls `load_heck_relay()` and asserts that the loader's active-mask param vector matches the values from `ForceField.from_mm3_fld(...) + freeze_standard_params(...)` (no Seminario) within `atol=1e-12`. Verified locally: passes after the fix; would fail loudly if any future change reintroduces the Seminario re-estimation on the OPT block.
